### DISCUSSION TITLE
Show & Tell — Phase 35.5 QuantumClassicalOrchestrator

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#736] [Show and tell]
### Architecture Overview

The QuantumClassicalOrchestrator is the capstone component of Phase 35, providing a unified pipeline that coordinates quantum circuit construction (35.1), variational optimization (35.2), feature map computation (35.3), and entanglement management (35.4) into a seamless hybrid computing framework. It handles qubit allocation, circuit routing, benchmark execution, and quantum advantage estimation.

```
┌─────────────────────────────────────────────────────────────────────┐
│                  QuantumClassicalOrchestrator                       │
│                                                                     │
│  ┌─────────────┐  ┌──────────────┐  ┌─────────────┐  ┌──────────┐  │
│  │ Classical    │  │ Quantum      │  │ Feature     │  │ Entangle │  │
│  │ Neural Net  │  │ Circuit      │  │ Map         │  │ Manager  │  │
│  │ Layers      │  │ Builder(35.1)│  │ (35.3)      │  │ (35.4)   │  │
│  └──────┬──────┘  └──────┬───────┘  └──────┬──────┘  └────┬─────┘  │
│         │                │                  │               │        │
│  ┌──────▼────────────────▼──────────────────▼───────────────▼─────┐  │
│  │                 Hybrid Execution Engine                        │  │
│  │                                                               │  │
│  │  ┌──────────────┐  ┌───────────────┐  ┌────────────────────┐  │  │
│  │  │ Qubit        │  │ Circuit       │  │ Variational        │  │  │
│  │  │ Allocator    │  │ Router        │  │ Optimizer (35.2)   │  │  │
│  │  └──────────────┘  └───────────────┘  └────────────────────┘  │  │
│  └───────────────────────────┬───────────────────────────────────┘  │
│                              │                                      │
│  ┌───────────────────────────▼───────────────────────────────────┐  │
│  │                 Results & Analysis Layer                       │  │
│  │                                                               │  │
│  │  ┌──────────────┐  ┌───────────────┐  ┌────────────────────┐  │  │
│  │  │ Benchmark    │  │ Advantage     │  │ Convergence        │  │  │
│  │  │ Suite        │  │ Estimator     │  │ Dashboard          │  │  │
│  │  └──────────────┘  └───────────────┘  └────────────────────┘  │  │
│  └───────────────────────────────────────────────────────────────┘  │
└─────────────────────────────────────────────────────────────────────┘
```

### Qubit Allocation Algorithm

The allocator maps logical qubits to physical qubits on the target hardware:

```python
class QubitAllocator:
    def allocate(self, circuit: QuantumCircuit, 
                 coupling_map: CouplingMap) -> Dict[int, int]:
        """Map logical to physical qubits minimizing SWAP overhead."""
        # Build interaction graph from circuit
        interaction_graph = self._build_interaction_graph(circuit)
        
        # Score each initial mapping using expected SWAP count
        best_mapping = None
        best_score = float('inf')
        
        # SABRE-style bidirectional search (Li et al. 2019)
        for trial in range(self.num_trials):
            # Random initial mapping
            mapping = self._random_mapping(circuit.num_qubits, coupling_map)
            
            # Forward pass: route circuit, count SWAPs
            fwd_swaps = self._route_forward(circuit, mapping, coupling_map)
            
            # Backward pass: route reversed circuit with final mapping
            rev_mapping = self._get_final_mapping(mapping, fwd_swaps)
            rev_swaps = self._route_forward(circuit.reverse(), rev_mapping, coupling_map)
            
            # Use backward pass result as improved initial mapping
            score = len(fwd_swaps) + len(rev_swaps)
            if score < best_score:
                best_score = score
                best_mapping = self._get_final_mapping(rev_mapping, rev_swaps)
        
        return best_mapping
```

### Circuit Routing with SWAP Insertion

For two-qubit gates on non-adjacent physical qubits, SWAP gates are inserted along the shortest coupling path:

```
Logical circuit:         Physical circuit (linear topology):
q0 --*--                 p0 --SWAP-- --*--
     |                          |      |
q3 --X--                 p1 --SWAP-- --+--
                                |      |
                         p2 ------- ---+--
                                       |
                         p3 ---------- X--
```

Each SWAP decomposes into 3 CNOTs, adding depth and noise. The router minimizes total SWAPs using a lookahead heuristic that considers the next k gates in the circuit queue.

**Metrics reported:**
- Number of inserted SWAPs
- Depth overhead (original vs routed circuit depth)
- Estimated fidelity impact (product of SWAP gate fidelities)
- Critical path analysis (longest qubit idle time)

### Hybrid Backpropagation

For quantum layers embedded in classical neural networks, the orchestrator supports end-to-end gradient computation:

```
Forward pass:
  x -> [Classical NN] -> theta -> [Quantum Circuit Q(theta)] -> <O> -> [Classical NN] -> y

Backward pass:
  dL/dy -> [Classical backprop] -> dL/d<O> -> [Parameter-shift rule] -> dL/dtheta -> [Classical backprop] -> dL/dx
```

```python
class HybridLayer(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inputs, weights, quantum_circuit, backend):
        ctx.save_for_backward(inputs, weights)
        ctx.quantum_circuit = quantum_circuit
        ctx.backend = backend
        expectation = backend.execute(quantum_circuit, weights)
        return expectation

    @staticmethod
    def backward(ctx, grad_output):
        inputs, weights = ctx.saved_tensors
        # Parameter-shift gradient for quantum parameters
        gradients = []
        for i in range(len(weights)):
            shifted_plus = weights.clone()
            shifted_plus[i] += np.pi / 2
            shifted_minus = weights.clone()
            shifted_minus[i] -= np.pi / 2
            grad_i = (ctx.backend.execute(ctx.quantum_circuit, shifted_plus)
                     - ctx.backend.execute(ctx.quantum_circuit, shifted_minus)) / 2
            gradients.append(grad_i)
        grad_weights = grad_output * torch.stack(gradients)
        return None, grad_weights, None, None
```

### Benchmark Suite Design

The orchestrator includes standardized benchmarks for evaluating quantum components:

| Benchmark | Problem | Qubits | Metric | Classical Baseline |
|-----------|---------|--------|--------|--------------------|
| Max-Cut QAOA | Graph optimization | 8-20 | Approximation ratio | Goemans-Williamson (0.878) |
| H2 VQE | Molecular energy | 4 | Chemical accuracy | FCI exact |
| LiH VQE | Molecular energy | 12 | Chemical accuracy | CCSD(T) |
| Quantum Kernel SVM | Classification | 5-15 | Test accuracy | RBF SVM |
| Random Circuit Sampling | Sampling | 10-30 | XEB fidelity | Brute force |
| Bernstein-Vazirani | Oracle | 5-20 | Success probability | Classical query complexity |

Each benchmark runs on both quantum and classical backends, with statistical comparison using paired t-tests and bootstrap confidence intervals.

### Advantage Estimation Methodology

Quantum advantage is assessed across three dimensions:

1. **Computational advantage**: Compare wall-clock time for quantum vs classical solution at equivalent quality. Account for compilation, communication, and measurement overhead.

2. **Solution quality advantage**: For optimization problems, compare objective values at fixed time budgets. Report advantage ratio = quantum_objective / classical_objective.

3. **Sample complexity advantage**: For learning tasks, compare test accuracy as a function of training set size. Quantum advantage manifests as steeper learning curves.

The estimator produces a structured report:
```python
@dataclass
class AdvantageReport:
    problem: str
    quantum_time: float      # seconds
    classical_time: float    # seconds
    quantum_quality: float   # problem-specific metric
    classical_quality: float
    speedup_ratio: float     # classical_time / quantum_time
    quality_ratio: float     # quantum_quality / classical_quality
    confidence_interval: Tuple[float, float]  # 95% CI for speedup
    verdict: str             # "advantage", "parity", "disadvantage"
    caveats: List[str]       # e.g., "simulator only", "small instance"
```

### Integration with Prior Phases

The orchestrator connects quantum capabilities to existing ASI-Build modules:
- **Phase 20 (Reasoning)**: Quantum optimization for combinatorial inference problems
- **Phase 22 (Creative Intelligence)**: Quantum sampling for stochastic generation
- **Phase 34 (Neuromorphic)**: Spiking neural networks as classical co-processors
- **Phase 33 (Continual Learning)**: Parameter persistence across quantum experiments

### Open Questions

1. How do we handle the latency of cloud quantum hardware in real-time pipelines?
2. What is the minimum problem size where quantum advantage becomes meaningful?
3. Should we implement quantum error correction codes or focus on error mitigation for NISQ?
4. How do we version and reproduce quantum experiments given hardware calibration drift?
